src/software/embedded/motor_controller/stspin_motor_controller.cpp

```diff
@@ -0,0 +1,239 @@
+    {
+        sendAndReceiveFrame(motor, StSpinOpcode::ACK_FAULTS);
+    }
+
+    for (const MotorIndex& motor : reflective_enum::values<MotorIndex>())
+    {
+        sendAndReceiveFrame(motor, StSpinOpcode::START_MOTOR);
+        checkDriverFault(motor);
+        readThenWriteVelocity(motor, 0);
+    }
+}
+
+void StSpinMotorController::reset()
+{
+    reset_gpio_->setValue(GpioState::LOW);
+}
+
+MotorFaultIndicator StSpinMotorController::checkDriverFault(const MotorIndex& motor)
+{
+    bool drive_enabled = true;
+    std::unordered_set<TbotsProto::MotorFault> motor_faults;
+
+    sendAndReceiveFrame(motor, StSpinOpcode::GET_FAULT);
+
+    // We must send a SPI_NOOP in order to clock out the response
+    // to the GET_FAULT request.
+    const uint16_t faults = sendAndReceiveFrame(motor, StSpinOpcode::SPI_NOOP);
+
+    if (faults != 0)
+    {
+        LOG(WARNING) << "======= Faults For Motor " << motor << "=======";
+    }
+
+    if (faults & static_cast<uint16_t>(StSpinFaultCode::DURATION))
+    {
+        LOG(WARNING) << "DURATION: FOC rate too high";
+        motor_faults.insert(TbotsProto::MotorFault::DURATION);
+        drive_enabled = false;
+    }
+
+    if (faults & static_cast<uint16_t>(StSpinFaultCode::OVER_VOLT))
+    {
+        LOG(WARNING) << "OVER_VOLT: Software overvoltage";
+        motor_faults.insert(TbotsProto::MotorFault::OVER_VOLT);
+        drive_enabled = false;
+    }
+
+    if (faults & static_cast<uint16_t>(StSpinFaultCode::UNDER_VOLT))
+    {
+        LOG(WARNING) << "UNDER_VOLT: Software undervoltage";
+        motor_faults.insert(TbotsProto::MotorFault::UNDER_VOLT);
+        drive_enabled = false;
+    }
+
+    if (faults & static_cast<uint16_t>(StSpinFaultCode::OVER_TEMP))
+    {
+        LOG(WARNING) << "OVER_TEMP: Software over temperature";
+        motor_faults.insert(TbotsProto::MotorFault::OVER_TEMP);
+        drive_enabled = false;
+    }
+
+    if (faults & static_cast<uint16_t>(StSpinFaultCode::START_UP))
+    {
+        LOG(WARNING) << "START_UP: Start up failed";
+        motor_faults.insert(TbotsProto::MotorFault::START_UP);
+        drive_enabled = false;
+    }
+
+    if (faults & static_cast<uint16_t>(StSpinFaultCode::SPEED_FDBK))
+    {
+        LOG(WARNING) << "SPEED_FDBK: Speed feedback fault";
+        motor_faults.insert(TbotsProto::MotorFault::SPEED_FDBK);
+        drive_enabled = false;
+    }
+
+    if (faults & static_cast<uint16_t>(StSpinFaultCode::OVER_CURR))
+    {
+        LOG(WARNING) << "OVER_CURR: Software overcurrent";
+        motor_faults.insert(TbotsProto::MotorFault::OVER_CURR);
+        drive_enabled = false;
+    }
+
+    if (faults & static_cast<uint16_t>(StSpinFaultCode::SW_ERROR))
+    {
+        LOG(WARNING) << "SW_ERROR: Software error";
+        motor_faults.insert(TbotsProto::MotorFault::SW_ERROR);
+        drive_enabled = false;
+    }
+
+    if (faults & static_cast<uint16_t>(StSpinFaultCode::SAMPLE_FAULT))
```

Comment:
What's a sample fault?

Reply:
Not sure, but I saw it in the firmware. Maybe it's used to test whether faults were being raised correctly

https://github.com/UBC-Thunderbots/MDV6_STSPIN/blob/8617178e99d35223f5afe659135daf382225e737/common/Inc/types.h#L42